### ISSUE TITLE
Refactor of the extra-valid-files cluster's task

### ISSUE THREAD

**Issue by AdriiiPRodri** [wazuh/wazuh#10123]
|Wazuh version|Component|
|---|---|
| 4.2.0+ | Wazuh cluster |

This issue has the objective of refactoring the cluster integrity sync task. Currently, we can separate this task in two subtasks.

On the one hand, we have a synchronization task that goes from the worker nodes to the master node, this task follows the same flow as the agent-info task. That is, permission is requested and then, once the master node has granted permission, the task is performed.

On the other hand, we have the extra-valid-files synchronization task. In this task, the worker node sends all the files requested by the master node.

The problem is that this last task is initialized once the master node sends the command to the worker node. At this moment, the worker node starts its task to finally send all the files to the master node. Right now, since we do not have any control mechanism for this task, it can happen that the master node requests the synchronization of extra-valid-files, the worker sends them and before the master node finishes applying the changes, the master itself sends again the request to synchronize the extra-valid-files and the worker sends them again. There is an overlap between the extra-valid-files tasks.

In this issue we want to refactor this part, making that this task cannot be initialized again if the master node has not finished processing the files, so we will follow the same scheme as the other tasks. We want the extra-valid-files task not to be a separate task in the worker.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

**Comment by Selutario** [comment]
# Issue update

Reviewing the time graphs for the integrity sync task, I found times that seemed very high, bordering on 120 seconds per synchronization in a cluster with 25 workers and 50,000 agents:
![image](https://user-images.githubusercontent.com/23361101/133992390-67f65de3-c39f-4657-baef-3d8cb9cfc4d0.png)

This was especially strange when compared to graphs from other tests with an identical configuration (25 workers and 50,000 agents), but that did not include the changes carried out in this development. In this case it seemed that most integrity syncs were usually below 35 seconds:
![image](https://user-images.githubusercontent.com/23361101/133992447-26e367dd-fd18-415f-89f7-f4dbe8eff74c.png)

After researching, there are two main reasons that explain the big differences between these two graphs:

### 1. Some logs were never printed before the refactor
- [Master's cluster.log before refactoring.](https://github.com/wazuh/wazuh/files/7195332/cluster.log)
- [Master's cluster.log after refactoring.](https://github.com/wazuh/wazuh/files/7195335/cluster.log)

If we look for the following string in the master's `cluster.log` before doing this refactor, we find 1238 matches.
```
[Integrity sync] Files to create in worker:
```

Instead, if we search for `[Integrity sync] Finished in`, only 603 matches are found even though the number of matches should be identical to the previous one (1238). It means that there are many `Finished in` logs that have not been printed, and the reason is the following:

If the master is waiting for extra-valid files, it does not print the `Finished in` log until it has received and processed those files. However, if the master had already received the files in a previous iteration but hadn't finished processing them yet, permission to send more extra-valid files wouldn't be granted to the worker, so the master would never print the log. Meanwhile, the worker would continue to synchronize other non-extra-valid files (missing, shared, extra) each 9 seconds, leading to many synchronizations whose duration is not logged:
```
2021/09/13 15:52:47 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity check] Starting. Received metadata of 15530 files.
2021/09/13 15:52:47 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity check] Finished in 10.927s. Sync required.
2021/09/13 15:52:47 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Starting.
2021/09/13 15:52:47 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Files to create in worker: 1386 | Files to update in worker: 1 | Files to delete in worker: 0 | Files to receive: 289
...
2021/09/13 15:53:19 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Starting.
2021/09/13 15:53:19 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Files to create in worker: 1027 | Files to update in worker: 1 | Files to delete in worker: 0 | Files to receive: 239
```
In the log above it can be seen that a new `Integrity sync` is started at `15:53:19`, but the `[Integrity sync] Finished in Xs` of the previous synchronization (`15:52:47`) was never printed. This was what was happening in the worker in the meantime:
```
2021/09/13 15:52:23 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Starting.
2021/09/13 15:52:23 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Files to create: 626 | Files to update: 1 | Files to delete: 0 | Files to send: 352
2021/09/13 15:52:24 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Master requires some worker files.
2021/09/13 15:52:24 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Starting sending extra valid files to master.
2021/09/13 15:52:25 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Finished in 1.370s. Master didn't grant permission to synchronize.
2021/09/13 15:52:25 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Finished sending extra valid files in 1.472s.
2021/09/13 15:52:25 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B43_manager_18] [Integrity sync] Finished in 1.861s.
```
As `Master didn't grant permission to synchronize.` extra valid files, the `Finished in` log would never be printed on his side. The longer tasks were not logged and it could seem that the synchronization was faster than it actually was.

### 2. Fewer syncs with more files
After the changes in this issue and as we have already explained in other updates, now the worker cannot start a new synchronization until the previous one has completely finished (including extra-valid). 

This results in fewer iterations, but each with more files, which can also lead to longer times. The cluster may seem to be slower because of this, but in reality, it will end up taking the same or even less time to complete. For example, this is what the setup phase took for each environment:
- **Before refactoring:** 
  - First sync iteration started at `15:47:26`.
  - Last sync iteration ended at `16:03:07`.
  - `1238` syncs took place during those **16 minutes**.
- **After refactoring:**
  - First sync iteration started at `10:00:10`.
  - Last sync iteration ended at `10:14:03`.
  - `627` syncs took place during those **14 minutes**.

There were half the iterations (627 vs 1238) but the same number of files ended up syncing in less time (14 vs 16 minutes).

Regards,
Selu.

**Comment by Selutario** [comment]
# Issue update

## Problem description
Right now, the `extra-valid` processing task on the master contains an `await asyncio.sleep(0)` inside a loop where all received files are iterated. This helps to relinquish control when other tasks come to the Asyncio queue (DAPI, agent-info, etc.) as we explained in other updates.
https://github.com/wazuh/wazuh/blob/bb1449a3bd89644ad8aed335439ceb6f94061555/framework/wazuh/core/cluster/master.py#L830

This can cause a similar problem to the one that Agent-info sync was suffering before the refactor that was carried out in [Refactor agent-info task to send all chunks from workers in just 1 task iteration #8056](https://github.com/wazuh/wazuh/issues/8056). 

The problem consisted in, when relinquishing the execution control (using await) after processing each chunk (or file in this case), if the environment is loaded and has chunks/files of other tasks, they are interspersed with each other and the task can be very delayed. For example, if we have 3 files to be processed coming from 3 different nodes, it would behave as shown below:
```
- Process file 0 from worker1
- Process file 0 from worker2
- Process file 0 from worker3
- Process file 1 from worker1
- Process file 1 from worker2
- Process file 1 from worker3
- Process file 2 from worker1
- Process file 2 from worker2
- Process file 2 from worker3
```

This causes the tasks of the 3 workers to finish practically at the same time. Therefore, the time it takes for the extra-valid synchronization to finish in each worker will be the equivalent of synchronizing all the files from all the workers.

## Proposed solution

Using await so many times to relinquish control can backfire. For this reason, a possible solution is to only use it every n files processed:
```PYTHON
                        processed_files += 1
                        if processed_files % max(
                                1, self.cluster_items['intervals']['master']['ev_files_processed_before_await']) == 0:
                            await asyncio.sleep(0)
```

In this way, in those cases in which there are less than `ev_files_processed_before_await` files to process, it would be carried out at once and the task would finish faster. On the other hand, when there are more than `ev_files_processed_before_await` files, the number of times that execution control is transferred will be greatly reduced.

## Results

We can compare two equal environments (50000 agents, 25 workers), but applying this solution to only one. This is the Integrity sync graph of the master without this improvement:
![image](https://user-images.githubusercontent.com/23361101/133992390-67f65de3-c39f-4657-baef-3d8cb9cfc4d0.png)

And this is after the proposed solution:
![image](https://user-images.githubusercontent.com/23361101/134142383-d9147e06-4381-40fc-ae8b-1891cc284f4b.png)

The set value for `ev_files_processed_before_await` was 100, that is, an `await asyncio.sleep(0)` is performed every 100 files processed. Using this configuration, it can be seen a considerable reduction in the time it takes for the tasks to complete. The maximum before the upgrade was close to 120 seconds, while it is now 50s. Furthermore, the mean (20.55s vs 12.47s) and median (12.44s vs 6.30s) have also decreased a lot.

However, this makes the distributed API requests slower, causing timeout errors:
```
2021/09/20 11:35:36 INFO: wazuh 172.31.32.158 "GET /security/user/authenticate" with parameters {} and body {} done in 0.546s: 200
2021/09/20 11:36:10 ERROR: Unhandled exception: Expecting value: line 1 column 1 (char 0)
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 157, in distribute_function
    response = await self.forward_request()
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 514, in forward_request
    response = await asyncio.shield(asyncio.gather(*[forward(node) for node in cleaned_valid_nodes]))
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 423, in forward
    result = json.loads(await client.execute(b'dapi_fwd',
  File "/var/ossec/framework/python/lib/python3.9/json/__init__.py", line 359, in loads
    return cls(**kw).decode(s)
  File "/var/ossec/framework/python/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/var/ossec/framework/python/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2021/09/20 11:36:10 INFO: wazuh 172.31.32.158 "PUT /security/config" with parameters {} and body {"auth_token_exp_timeout": 999999} done in 34.930s: 500
```

This might be solved by using a lower `ev_files_processed_before_await` value (50 or 30), although this would also reduce the improvement in integrity sync. 

The definitive solution will be to use different processes for the DAPI and the cluster tasks, as reported at [Improve distributed cluster and API performance: multi-process #8251](https://github.com/wazuh/wazuh/issues/8251). Until then and to avoid further problems, we have decided not to include this change, only the refactoring of extra-valid.

Regards.

**Comment by Selutario** [comment]
# Issue update

We have discarded one of the changes that were carried out in this issue. 

We made the workers not show the log `[Integrity sync] Finished in Xs` until the master finished processing the extra-valid files. Now, the workers show `[Integrity sync] Finished in Xs` again as soon as they have sent those files, without waiting for the master to confirm that it has already processed them.

This is due to the fact that the operations carried out on the master are different from those of the workers, so it does not make sense that they log the same duration. Also, it can be more useful for debugging if each node shows the actual time it takes for it to finish.

**Comment by Selutario** [comment]
# Issue update

[artifacts.zip](https://github.com/wazuh/wazuh/files/7218600/artifacts.zip)

We set up a new environment with 25 workers and 50000 agents after the changed performed in [this commit](https://github.com/wazuh/wazuh/pull/10183/commits/124209517d3c93c29d3e8cfcee627269617d666e) (#L201):
https://github.com/wazuh/wazuh/blob/df917592224992c22f8934082766f7b8c832fa05/framework/wazuh/core/cluster/common.py#L188-L201

Strange results have been obtained. On the one hand, the tasks have taken exaggeratedly longer (400+ seconds sometimes) than in previous builds for the same size:
![image](https://user-images.githubusercontent.com/23361101/134527323-442da66f-c312-426d-9d28-5b7a2cda5ebd.png)

On the other hand, multiple errors can be seen in all nodes due to timeouts:
```
2021/09/22 13:30:45 ERROR: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_15] [Main] Error sending files information: Error 3016 - Received an error response: b'Error sending request: timeout expired.'
```

It seems that on some (or many) occasions, an integrity check is being started while the previous integrity sync has not finished yet, which should never happen as permission should be denied. For instance, this can be seen in the worker 25:
```
2021/09/22 13:20:20 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Permission to synchronize granted.
2021/09/22 13:20:20 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Starting.
2021/09/22 13:20:20 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Permission to synchronize granted.
2021/09/22 13:20:20 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Starting.
2021/09/22 13:20:20 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Obtained 13 chunks of data in 0.045s.
2021/09/22 13:20:21 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Compressing 'files_metadata.json'.
2021/09/22 13:20:21 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Sending zip file to master.
2021/09/22 13:20:30 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] All chunks sent.
2021/09/22 13:20:30 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: 'b'syn_m_a_e''
2021/09/22 13:20:30 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Finished in 12.177s (13 chunks updated).
2021/09/22 13:20:31 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Zip file sent to master.
2021/09/22 13:20:31 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: 'b'syn_m_c''
2021/09/22 13:20:31 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Finished in 14.465s. Sync required.
2021/09/22 13:20:34 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: 'b'new_file''
2021/09/22 13:20:45 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: 'b'file_upd''
2021/09/22 13:20:45 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Starting.
2021/09/22 13:20:46 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Permission to synchronize granted.
2021/09/22 13:20:46 INFO: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Starting.
2021/09/22 13:20:46 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Agent-info sync] Obtained 14 chunks of data in 0.043s.
2021/09/22 13:20:46 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Compressing 'files_metadata.json'.
2021/09/22 13:20:46 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Integrity check] Sending zip file to master.
```

Integrity check started at `13:20:20` and then again at `13:20:45`, but no Integrity sync was performed in between. 

The problem is that at the moment and after some hours of debugging, I have not been able to find the root of the problem, so I cannot determine if it is a consequence of the change in https://github.com/wazuh/wazuh/pull/10183/commits/124209517d3c93c29d3e8cfcee627269617d666e or if it is a race condition that is not related to said commit and that could happen on more occasions.

In the master we can also find strange situations. For example:
```
2021/09/22 13:24:14 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: b'syn_i_w_m' 
2021/09/22 13:24:14 DEBUG: [Worker CLUSTER-Workload_benchmarks_metrics_B48_manager_25] [Main] Command received: b'syn_e_w_m_e'
```
In the same second (`13:24:14`), the `syn_i_w_m` command and the `syn_e_w_m_e` command have been received.
- `syn_i_w_m` indicates that an integrity check/sync is starting and sets the `sync_integrity_free` flag to false so the worker has no permission to start new synchronizations. 
- `syn_e_w_m_e` indicates that the worker has finished sending extra-valid files and the master can process them. 

When the master finishes processing the received extra-valid files, the `sync_integrity_free` flag will be True again so new synchronizations can occur, even if one is already in progress. But a new sync shouldn't be able to start (it is, `syn_i_w_m` should have not been received) since the worker is supposed to be unable to start new synchronizations until all extra-valid files are processed. 

A race condition must have caused a new synchronization to be allowed to start when the previous one has not finished, and this causes a chain reaction. I will continue investigating the origin.

**Comment by Selutario** [comment]
## Issue update

I have reviewed the results of a new workload test (50000 agents, 25 workers) after the changes performed in https://github.com/wazuh/wazuh/pull/10183/commits/af26462d18b712e2a6bb117df32d86618c6e1384.

I have not found duplicate or overlapping Integrity tasks in the master logs. In the same way, the workers' logs follow the expected order, without skipping any step (Integrity check -> Integrity sync). This is also reflected in the Integrity sync graphs, for example in the master's:

![image](https://user-images.githubusercontent.com/23361101/134932789-15b0a7f8-745e-456c-b736-3d35380c4e11.png)

In some cases peaks of 70 seconds are reached, but they can be considered normal given the size of the cluster. In addition, those results are very far from the times that were obtained with the race condition described in the Issues updates of these last days, where some synchronizations could take more than 400 seconds: https://github.com/wazuh/wazuh/issues/10123#issuecomment-925911618

As a summary, everything seems correct now.
